week3/ml_decision_tree_sample/train.py

```diff
@@ -0,0 +1,36 @@
+import hydra
+from omegaconf import DictConfig
+import sklearn
+import wandb
+from sklearn.ensemble import RandomForestClassifier
+from prepare_data import prepare_data
+
+@hydra.main(version_base=None, config_path="conf", config_name="train")
+def classifier_visualizations(cfg: DictConfig):
+    prepared_data = prepare_data(cfg.path.train_file)
+    parange = cfg.range_parametrs.common_range
+    wandb.init(project="ml_week3", entity="liunomyr")
+    wandb.define_metric("train/step")
+    X_train = prepared_data[0]['x']
+    X_test = prepared_data[1]['x']
+    y_train = prepared_data[0]['y']
+    y_test = prepared_data[1]['y']
+    for max_depth in range(1, parange):
+        for min_samples_leaf in range(1, parange, 2):
+            for min_samples_split in range(2, parange, 3):
+                for n_estimators in range(2, parange, 4):
+                    clf = RandomForestClassifier(max_depth=max_depth, min_samples_leaf=min_samples_leaf,
+                                                 min_samples_split=min_samples_split,
+                                                 n_estimators=n_estimators, random_state=0)
+                    clf.fit(X_train, y_train)
+                    y_pred = clf.predict(X_test)
+                    if sklearn.metrics.precision_score(y_test, y_pred) > 0.2:
+                        wandb.log({
+                            "precision": sklearn.metrics.precision_score(y_test, y_pred, zero_division=0),
+                            "recall": sklearn.metrics.recall_score(y_test, y_pred, zero_division=0),
+                            "f1_score": sklearn.metrics.f1_score(y_test, y_pred, zero_division=0),
+                            "accuracy": sklearn.metrics.accuracy_score(y_test, y_pred),
+                            "train data": {"max_depth": max_depth, "min_samples_leaf": min_samples_leaf,
+                                           "min_samples_split": min_samples_split,
+                                           "n_estimators": n_estimators}
+                        })
```

Comment:
you have 5 nested structures here, more than 3 is already hard to read, let's move some code into separate functions

Reply:
changed

Reply:
It will not see modules otherwise